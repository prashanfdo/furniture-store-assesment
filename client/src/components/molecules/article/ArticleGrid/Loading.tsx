import ContentLoader from "react-content-loader";

const LoadingArticle = () => {
  return (
    <ContentLoader
      data-testid="loading"
      speed={2}
      width={"100%"}
      height={300}
      backgroundColor="#fff"
      foregroundColor="#f0f0f0"
    >
      <rect x="0" y="0" rx="5" ry="5" width="100%" height="300" />
    </ContentLoader>
  );
};
const Loading = () => {
  return (
    <>
      <LoadingArticle />
      <LoadingArticle />
      <LoadingArticle />
    </>
  );
};

export default Loading;
