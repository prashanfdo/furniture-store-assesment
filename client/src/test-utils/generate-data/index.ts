export * from "./article";
export * from "./category";
export * from "./search";
