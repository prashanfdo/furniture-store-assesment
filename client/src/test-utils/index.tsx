import { render, RenderOptions } from "@testing-library/react";
import fs from "fs";
import React, { FC, ReactElement } from "react";

import { BrowserRouter } from "react-router-dom";
import user from "@testing-library/user-event";

const wrapper: FC<{ children: React.ReactNode }> = ({ children }) => {
  return <BrowserRouter>{children}</BrowserRouter>;
};

const customRender = (ui: ReactElement, options?: Omit<RenderOptions, "wrapper">) => {
  const view = render(ui, { wrapper, ...options });

  const style = document.createElement("style");
  style.innerHTML = fs.readFileSync("src/index.out.css", "utf8");
  document.head.appendChild(style);

  return view;
};
export * from "@testing-library/react";

export { faker } from "@faker-js/faker";
export * from "./generate-data";

export { customRender as render };
export { user };
