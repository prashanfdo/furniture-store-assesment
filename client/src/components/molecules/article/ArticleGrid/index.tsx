import { Article } from "types";
import { ArticleThumb } from "components/molecules";
import Loading from "./Loading";

type ArticleGridProps = Pick<React.ComponentProps<typeof ArticleThumb>, "onAddtoCart"> & {
  products: Article[];
  loading: boolean;
};
export default function ArticleGrid({ loading, products, onAddtoCart }: ArticleGridProps) {
  return (
    <div data-testid="product-grid" className="grid grid-cols-[repeat(auto-fill,_minmax(280px,_1fr))] gap-4">
      {loading && <Loading />}
      {/* using index as key since api doesn't returns an id and names are not unique to be used as key */}
      {products.map((product, index) => (
        <ArticleThumb key={index} product={product} onAddtoCart={onAddtoCart} />
      ))}
    </div>
  );
}
