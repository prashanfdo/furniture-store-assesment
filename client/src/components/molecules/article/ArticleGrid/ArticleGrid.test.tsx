import { createArticleList, render, screen, user } from "test-utils";
import ArticleGrid from "./index";

jest.mock("react-content-loader", () => () => <div data-testid="loading">Loading</div>);

const dataProducts: React.ComponentProps<typeof ArticleGrid>["products"] = createArticleList(10);

describe("ArticleGrid", () => {
  beforeEach(() => {
    user.setup();
  });

  it("should render", () => {
    const handleAddToCart = jest.fn();
    const { container } = render(<ArticleGrid products={dataProducts} onAddtoCart={handleAddToCart} loading={false} />);
    expect(container).toMatchSnapshot();
    const productGrid = screen.getByTestId("product-grid");
    expect(productGrid).toBeInTheDocument();
    expect(productGrid.childNodes).toHaveLength(dataProducts.length);
  });

  it("should invoke onAddtoCart callback", async () => {
    const handleAddToCart = jest.fn();
    render(<ArticleGrid products={dataProducts} onAddtoCart={handleAddToCart} loading={false} />);
    await user.click(screen.getAllByRole("button")[0]);
    expect(handleAddToCart).toHaveBeenCalledTimes(1);
  });
});
