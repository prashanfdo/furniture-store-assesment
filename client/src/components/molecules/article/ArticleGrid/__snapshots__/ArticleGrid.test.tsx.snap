// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ArticleGrid should render 1`] = `
<div>
  <div
    class="grid grid-cols-[repeat(auto-fill,_minmax(280px,_1fr))] gap-4"
    data-testid="product-grid"
  >
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Tasty Metal Sausages"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Tasty Metal Sausages
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          1.512,00 €
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Tasty Bronze Bike"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Tasty Bronze Bike
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          2.586,00 €
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Practical Plastic Chicken"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Practical Plastic Chicken
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          4.743,00 €
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Luxurious Cotton Gloves"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Luxurious Cotton Gloves
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          2.814,00 €
        </span>
        <span
          class="ml-4 text-sm line-through"
          data-testid="regular-price"
        >
          3.098,00 €
        </span>
        <span
          class="px-1 py-1 ml-2 text-sm text-white bg-red-500 rounded-lg"
          data-testid="discount-percent"
        >
          -
          9
          %
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Refined Wooden Chair"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Refined Wooden Chair
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          3.682,00 €
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Modern Cotton Towels"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Modern Cotton Towels
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          615,00 €
        </span>
        <span
          class="ml-4 text-sm line-through"
          data-testid="regular-price"
        >
          1.561,00 €
        </span>
        <span
          class="px-1 py-1 ml-2 text-sm text-white bg-red-500 rounded-lg"
          data-testid="discount-percent"
        >
          -
          61
          %
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Oriental Soft Fish"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Oriental Soft Fish
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          4.211,00 €
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Electronic Metal Hat"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Electronic Metal Hat
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          4.460,00 €
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Licensed Bronze Bacon"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Licensed Bronze Bacon
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          1.839,00 €
        </span>
        <span
          class="ml-4 text-sm line-through"
          data-testid="regular-price"
        >
          2.092,00 €
        </span>
        <span
          class="px-1 py-1 ml-2 text-sm text-white bg-red-500 rounded-lg"
          data-testid="discount-percent"
        >
          -
          12
          %
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
    <a
      class="block p-4 overflow-hidden bg-white border-none rounded-md decoration-transparent"
      href="/#product-page"
    >
      <div>
        <img
          alt="Refined Metal Computer"
          class="object-contain w-full h-96"
          loading="lazy"
          src="https://loremflickr.com/640/480"
        />
      </div>
      <h4
        class="overflow-hidden text-gray-900 text-ellipsis whitespace-nowrap"
      >
        Refined Metal Computer
      </h4>
      <div
        class="text-gray-600"
      >
        <span
          class="font-semibold text-red-600"
          data-testid="final-price"
        >
          2.684,00 €
        </span>
      </div>
      <button
        class="text-center border-solid border-gray-200 border block w-full p-2 rounded mt-2 cursor-pointer bg-white text-gray-600 hover:bg-[#7047eb] hover:text-white transition-all duration-100"
      >
        Add To Cart
      </button>
    </a>
  </div>
</div>
`;
